### DISCUSSION TITLE
🐗 Boar Adventure - Ideas

### DISCUSSION THREAD

**Discussion by Majorfi** [Rarity-Extended/rarity_extended_boars#1] [Dev & progress]
## Boar Adventure 🐗
Let's discuss the Boar Adventure
Contract : https://github.com/Rarity-Extended/extendedAdventures/blob/main/contracts/boarAdventure.sol

**Comment by Ceazor** [comment]
Just to make this a bit more interesting ..

```
function bonus_by_attr(uint _points, uint _summoner) internal view returns (uint points) {
        (,,,uint32 _int,uint32 _wis,uint32 _cha) = attr.ability_scores(_summoner);
        points = _points + ((_int + _wis + _cha) / 2);
```
        
The contracts will also consider this all. But the UI can ask.

"What would you like to do with the BOAR?"

>KILL IT (uses _str)

Attempt to get it to breed with other boars to increase the population and perhaps finds some goodies.

>RECALL WHAT YOU KNOW ABOUT BOAR (This will use your INTELLIGENCE)
>TRACK OTHER BOARS AND SEE WHAT YOU FIND (This will use your WISDOM)
>ATTEMPT TO COMMUNICATE WITH THE BOAR (This will use your CHARISMA)
>USE YOUR DEEP KNOWLEDGE OF ANIMALS TO TRY TO TAME IT (This will use yor and ANIMAL HANDLING skill)

**Comment by Ceazor** [comment]
I see that this block is multiplied by 2, 

```
function bonus_by_handle_animal(uint _points, uint _summoner) internal view returns (uint points) {
        uint8[36] memory _skills = skills.get_skills(_summoner);
        uint handle_animal = _skills[13]; //Handle animal
        points = _points + (handle_animal * 2);
```


but usually ANIMAL HANDLING bonus is added to that of the _cha bonuses since it's a charisma based skill. 

```
function bonus_by_handle_animal(uint _points, uint _summoner) internal view returns (uint points) {
        uint8[36] memory _skills = skills.get_skills(_summoner);
        uint handle_animal = _skills[13]; //Handle animal
        points = _points + (handle_animal + _cha);
```

not sure if this is right..

**Reply by TheAustrian1998** [reply]
Is coherent, i will add it

**Reply by TheAustrian1998** [reply]
Do you agree if it's like this:
`function bonus_by_handle_animal(uint _points, uint _summoner) internal view returns (uint points) {
          uint8[36] memory _skills = skills.get_skills(_summoner);
          uint handle_animal = _skills[13]; //Handle animal
          points = _points + (handle_animal * (_cha / 2));`

So, if your handle animal is zero: 0 * (x / 2) = 0.
But, if your handle animal is more than zero, it boost by your charism.

**Reply by Ceazor** [reply]
Im a little bit confused by 

```
function simulate_reproduce(uint _summoner) public view returns (uint reward) {
        uint _level = rm.level(_summoner);
        uint _class = rm.class(_summoner);
        reward = base_points_by_class(_class);
        reward = multiplier_points_by_level(reward, _level);
        reward = bonus_by_handle_animal(reward, _summoner);
        reward = bonus_by_attr(reward, _summoner);
```
this is going to make some classes better than others at this?
```
        reward = base_points_by_class(_class); 
```
this makes sense.. higher level get more.
```
        reward = multiplier_points_by_level(reward, _level); 
```
Are these IF/ELSE or are they SUMed? if SUMed we need to remove my idea of addign Char to bonus_by_handle_animal
```
        reward = bonus_by_handle_animal(reward, _summoner);
        reward = bonus_by_attr(reward, _summoner); 
```

**Reply by TheAustrian1998** [reply]
> this is going to make some classes better than others at this?

Yes

> Are these IF/ELSE or are they SUMed?

Yes, it is sumed, it accumulates all the bonuses in reward variable.

**Reply by Ceazor** [reply]
At least i learn.. so no CHA + in handle_animal

just make them straight summing

so.

Level 4 Bard with Wis 2 and AH 2will have 

reward = 4 + 4 + 2 + 2

That being said.. I would vote for change these and make them a lot less important 

```
  function base_points_by_class(uint _class) internal pure returns (uint points) {
        if (_class == 1) {
            points = 1;
        } else if (_class == 2) {
            points = 4;
        } else if (_class == 3) {
            points = 2;
        } else if (_class == 4) {
            points = 6;
        } else if (_class == 5) {
            points = 1;
        } else if (_class == 6) {
            points = 2;
        } else if (_class == 7) {
            points = 1;
        } else if (_class == 8) {
            points = 4;
        } else if (_class == 9) {
            points = 1;
        } else if (_class == 10) {
            points = 1;
        } else if (_class == 11) {
            points = 2;
        }
    }
```
only classes i would consider more able to handle_animal   are the druid and ranger.  and the wizard, perhaps.

**Reply by TheAustrian1998** [reply]
Okay, commit in a while. So, `points = _points + (handle_animal * 2);` this line keeps unmodified?

**Reply by Ceazor** [reply]
im not sure why you are * 2 actually.. just to make it more important?

**Reply by TheAustrian1998** [reply]
Yes, thus it produces a great boost

**Reply by Ceazor** [reply]
Great.. sound good...

**Comment by Majorfi** [comment]
## Part 1

### Dungeon Stats
Imo, we should make the dungeon harder than the current one. You must have a nice constitutions & strength to be able to kill a boar (much more hard than a rat). We need to test the correct combinaison here, but this could be something like this (note : not tested at all, but I can make a UI to test the fight)

```diff
+    int public constant dungeon_health = 16;
-    int public constant dungeon_health = 10;

+    int public constant dungeon_damage = 4;
-    int public constant dungeon_damage = 2;

+    int public constant dungeon_to_hit = 5;
-    int public constant dungeon_to_hit = 3;

+    int public constant dungeon_armor_class = 4;
_    int public constant dungeon_armor_class = 2;
```

### Boars Population
The initial boar population should no be `0`. Maybe something like `1000` ?
Moreover, to be able to reproduce a Boar, at least 2 boars should exists.

`A litter of piglets can be between 10 and 12 piglets` -> Do you want to increase the number of boars by `RANDOM 1 to 12` or is it too complicated ? If so, the killing boars should be a lot more interesting (or we will destroy the forest with an army of 100000000 boars haha) (not sure this is a good idea)

### Rewards
Perfect.
Just to be sure, the contract will be a minter for theses ERC20 right ? How should we decide who can mint or not (having this mint function accessible is really nice, we could add more contracts and allow the community to use that, this is brillant ! But we need to controle this to avoid a minter that will just mint and broke the system. Should our mutltisig be the owner ?)
When killing, you only go 1 reward among the 3 possibles ? Should we allow a random stuff to get more than just 1 type of reward (or more than just 1 meat for example) ?
IMO, the quantity should be partly random (cf the crafting contract). I find this sad that people can, for sure, know what kind of reward they will get.

### DAY Mechanisms
- With the `DAY` variable, we could imagine something to reduce the time between two actions with the Boars. For example, `1 DAY` is standard, but for 1 FTM you could rest and reduce this time to `0.5 DAY`. Or you could craft a `Meal`, with `4 berries + 1 meat` to reduce this time to `0.7 DAY`.


### Feats
In the same way as the skills, some feats could be used. All the feats can be [found here](https://www.d20srd.org/srd/feats.htm#featDescriptions), but here is a small selection of stuff, not too difficult (imo) to integrate :

> Alertness [General]
> Benefit
> You get a +2 bonus on all Listen checks and Spot checks.

> Animal Affinity [General]
> Benefit
> You get a +2 bonus on all Handle Animal checks and Ride checks.

> Endurance [General]
> Benefit
> You gain a +4 bonus on the following checks and saves: Swim checks made to resist nonlethal damage, Constitution checks made to continue running, Constitution checks made to avoid nonlethal damage from a forced march, Constitution checks made to hold your breath, Constitution checks made to avoid nonlethal damage from starvation or thirst, Fortitude saves made to avoid nonlethal damage from hot or cold environments, and Fortitude saves made to resist damage from suffocation. Also, you may sleep in light or medium armor without becoming fatigued.

> Toughness [General]
> Benefit
>You gain +3 hit points.


### Skills
The Handle_animal skill is perfect for this situation. Do we also need the [`Move Silently` skill](https://www.d20srd.org/srd/skills/moveSilently.htm) ? Bonus for both actions ? (not sure about this)

**Reply by TheAustrian1998** [reply]
> Do you want to increase the number of boars by RANDOM 1 to 12 

Yes, I can do that with the codex you gave me. To clarify, it is the amount of boars that will born when you reproduce?

> Should our mutltisig be the owner

Maybe, or I can add OpenZeppelin Admin Roles, so we can add multiple smart contracts as minter. _haha money printer go brrrr_

> Should we allow a random stuff to get more than just 1 type of reward (or more than just 1 meat for example) ?

Interesting mechanism, agree, we should add it.

> the quantity should be partly random (cf the crafting contract)

Agree with you, i made the mechanism "deterministic" because do not want to break the simulate functions. If we get off the road this, everything is more fun.

About the day mechanism, agree too. Maybe add some forest objects or crafts from rarity to increase loot or something.

**Reply by Ceazor** [reply]
adding in forest objects would be great

[https://github.com/Rarity-Extended/RarityExtended/blob/main/utils/codex/items_dungeon_theForest.json](url)

Wisdom bonus
"Antique ring"
"Cursed book" (curse)

INT bonus
"Cursed book"
"Ancient Prince Andre's Sword"

CHA Bonus
"Enchanted useless tool"

Animal Handling
"Dragon egg"
"Unknown animal eye"

**Reply by Majorfi** [reply]
About theses object, should they be burned after use? Maybe they should have a chance to break?

**Reply by TheAustrian1998** [reply]
@Major-Eth What do you think about getting rid off _simulate_ functions? I think they are needed for simulate the battle in UI. For that reason, i need your opinion

**Reply by Majorfi** [reply]
The function is not needed for the UI, but this can help people know if they will succeed this dungeon.
Could we just transform the `simulate` in a [`Yes`/`No`] situation ? Yes if you could Win the fight, no otherwise. Nothing about loot, just success/failure ?

**Reply by TheAustrian1998** [reply]
This makes everything so much easy, thanks. So, we only need a simulate kill function, because reproduce is always successfully, (only change your reward).

**Reply by Majorfi** [reply]
> Do you want to increase the number of boars by RANDOM 1 to 12
> Yes, I can do that with the codex you gave me. To clarify, it is the amount of boars that will born when you reproduce?

Exactly, the amount of boars could be a random number. But, if we do this, we will have to get a counter mecanism to avoid TOO MANY boars. This mecanism could be a reduction of loots for the `Handle` part, and an increase for the `Kill` part.  
Let's say the expected number is 1000 boars, then the loot amount could be something like this : 
```diff
function simulate_reproduce(uint _summoner) public view returns (uint reward) {
        uint _level = rm.level(_summoner);
        uint _class = rm.class(_summoner);
        reward = base_points_by_class(_class);
        reward = multiplier_points_by_level(reward, _level);
        reward = bonus_by_handle_animal(reward, _summoner);
        reward = bonus_by_attr(reward, _summoner);
+      reward = reward * (EXPECTED_BOARS_AKA_1000 / CURRENT_BOARS_COUNT_AKA_2000)
}
```
And so reward become reward / 2 -> `1000/2000 = 0.5`, and the greater CURRENT_BOARS_COUNT is, the less reward there is.
And we could do the same thing for the kill mecanism:
```
reward * (CURRENT_BOARS_COUNT_AKA_500 / EXPECTED_BOARS_AKA_1000)
```

**Reply by TheAustrian1998** [reply]
Wow, very interesting mechanism. I will add it.

**Comment by Ceazor** [comment]
is this a typo? [13]?

```
function bonus_by_handle_animal(uint _points, uint _summoner) internal view returns (uint points) {
uint8[36] memory _skills = skills.get_skills(_summoner);
uint handle_animal = _skills[13]; //Handle animal
points = _points + (handle_animal * 2);
```
![image](https://user-images.githubusercontent.com/73511897/135577488-6163c32d-cb6b-4071-b2c9-2c0037b9739d.png)

**Reply by TheAustrian1998** [reply]
It isn't a typo. When you get skills array starts with zero, in the screenshot you provided it starts with one. Anyway, you make me doubt, i will double check.

**Reply by Majorfi** [reply]
Indeed, now that you say this I think you are right
